### ISSUE TITLE
Guidance: APM Instrumentation 

### ISSUE THREAD

**Issue by rsmets** [feathersjs/feathers#2123]
### Steps to reproduce

I would love to use New Relic to instrument my feathers app however it is not supported natively by New Relic APM. I know New Relic custom instrumentation is possible and I found a [feathers issue](https://github.com/feathersjs/feathers/issues/1117) from 11/30/2018 on the subject however the workaround example code provided by @daffl does not seem to work with recent versions of feathers given the `transaction` attribute on the ts type HookContex seems to be no longer. I am using feathers 4.5.2. 

I am curious what is the best way to instrument the app for use with an APM, in my case I specifically trying to use new relic but a general approach would suffice too, while not having a transaction attribute.  

Below is my attempted app.hooks.ts, a type script adapted version of  the example code authored by @daffl in the issue linked above.

```
import { HookContext } from '@feathersjs/feathers';
import { pick } from 'lodash';

import logger from './logger';
import { setDefaultHeaders } from './services/hooks/setDefaultHeaders';
import { downcaseHeaderKeys } from './services/hooks/downcaseHeaderKeys';
import { checkAdminKey } from './services/hooks/checkAdminKey';
import newrelic from 'newrelic';

function newRelicTransactionBefore (ctx: HookContext): Promise<Record<string, any>> {
  return new Promise((resolve, reject) => {
    newrelic.startWebTransaction(`${(ctx.method)} ${ctx.path}`, error => {
      if (error) {
        return reject(error);
      }

      // This doesn't work because no transaction attribute on HookContext =(
      ctx.transaction = newrelic.getTransaction();

      resolve(ctx);
    });
  });
}

function newRelicTransactionAfter (ctx: HookContext): HookContext {
  // This doesn't work because no transaction attribute on HookContext =(
  if (ctx.transaction) {
    ctx.transaction.end();
  }

  return ctx;
}

function logRequest (ctx: HookContext): void {
  const { path, method, id } = ctx;
  logger.info(`${path}#${method}${id ? ` id: ${id}` : ''}`);
}

function logError (ctx: HookContext): void {
  const { path, method, error } = ctx;
  const { name, code, message } = error;
  const rawInfo = pick(ctx, ['params', 'id', 'data']);
  const info = { ...rawInfo };

  if (info.params?.headers?.authorization) {
    info.params.headers.authorization = 'Bearer *****';
  }

  if (info.params.authToken) {
    info.params.authToken = '*****';
  }

  logger.warn(`Error in ${path}#${method}: name=${name} code=${code} message=${message}`, info);
}

export default {
  before: {
    all: [logRequest, newRelicTransactionBefore],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  after: {
    all: [newRelicTransactionAfter],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  error: {
    all: [logError],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  }
};
```

Thanks for your input and guidance!

**Comment by daffl** [comment]
You can add any property to the hook context. If TypeScript complains for some reason just cast the context to `any` or extend the hook context type.

**Comment by rsmets** [comment]
Yep, totally. Thank you for the suggestion. I have actually went ahead and done that. 

However, it seems due to the nature of the service hooks if one service calls another one, the first service request to "finish", which is the last called in the call graph (aka the nested call) the new relic transaction is ended with that one and does not capture the parent caller. 

This seems to be the down side of leveraging feathersjs without some sort of express endpoint sitting in front of it so instrumentation can happen more cleanly? Am I missing something? Feathersjs seems to be great at as a crud data interface service layer.

**Comment by daffl** [comment]
You can still do instrumentation with a normal Express middleware which will then only profile the actual HTTP requests. If you need the transaction everywhere, you could e.g. add it as `context.params.transaction` and pass it along in nested calls.

**Comment by rsmets** [comment]
Ahh interesting that adding the instrumentation to the express middleware would work. I was under the impression that feathersjs leveraged itself over the loopback ultimately sitting on top of express. Meaning that adding this to the express middle where would just result in the same behavior. Sounds like I assumed wrong?

Thanks for the suggest adding a `transaction` attribute to the `context.params`, makes sense. Probably a better way to handle than what I ultimately ended up with which was a transaction handler LRU type solution.

**Comment by rsmets** [comment]
Upon adding the instrumentation to the express middleware it indeed did just capture the original request, clearly my assumption was incorrect. Unfortunately, the data captured is not very useful (no total time to serve response or transaction trace bumping around service to service) aside from correctly creating a single transaction per api request. 

Upon trying the `context.params.transaction` suggestion it seems I would have to pass that param everywhere which seems arduous?

**Comment by DaddyWarbucks** [comment]
So a couple of things here.

>  I was under the impression that feathersjs leveraged itself over the loopback ultimately sitting on top of express

Feathers actually sits _under_ express and not _over_ express like most frameworks. Express is just a transport, similar to socket.io. A "transport" in this case is really any way to convert data to a feathers method. A Feathers "app" is really just a collection of services, and not necessarily even a REST framework. You could even create a Feathers app in a browser client that CRUD's on localtStorage. So, Express sits on _top_ of feathers and converts the req to a feathers service. You mentioned `loopback`...I think there may be some confusion there. Express !== Loopback

> Upon trying the context.params.transaction suggestion it seems I would have to pass that param everywhere which seems arduous?

A new context is created on every service call. Feathers does not magically "share" context between service calls. The issue you linked does not consider that and was, I believe, only setup for one service call. You have to pass things from service to service manually via params (and you are right, that is arduous). Again, you mentioned loopback, which does actually share context between requests. That is done via a node module called `async_hooks` (similar name, very different). There is actually a proposal to add this in the next feathers version: https://github.com/feathersjs/feathers/issues/2074. Passing things from service to service is a bit of a pain, but is also integral to how flexible Feathers is and what a Feathers "app" really is.

For us to implement the `async_hooks` (which is actually a total working example in that link, its really simple) easily, we need the new "feathers async hooks" to land in V5 (which are similar to Koa middleware) to stay in the same "execution context" for that to work. 

For now, the only way to really handle this is to manually pass the transaction around via params. You could maybe also create a `mixin`, or extend classes, to handle it.

**Comment by rsmets** [comment]
Thank you for clarifying the network interface @DaddyWarbucks, makes sense. I found myself reading the docs yesterday to solidify that. Coming from the re-incarnated [parse-server](https://github.com/parse-community/parse-server) world just made the (incorrect) query over loopback assumption. 

However, your proposal, https://github.com/feathersjs/feathers/issues/2074, is a timely one from my perspective! Will keep my eyes on it. Maybe could even lend a hand if I end up doing something similar for my own use case. 

Thank you for your response, thoughts, and work for the proposal for more streamlined service <-> info sharing. Sounds like that does somewhat go against the modularity of the feather's service design principals but there is just tremendous stream-lined usability value I would argue. 

I will probably close this issue in the next day or two if no one else chimes in, but I think I have a handle why this was so (surprisingly) difficult for me to accomplish. 

Side note: I am still super curious to hear how others have instrumented their feathers app for application observability (APM).

**Comment by DaddyWarbucks** [comment]
@rsmets I think most folks have integrated that type of thing manually (passing transactions via params). You can also checkout `feathers-profiler`.

**Comment by DaddyWarbucks** [comment]
@rsmets Has parse-sever seen another resurgence? I know it was open sourced a popular a few years back but I thought it kind of fizzled out. You mentioned "re-incarnated", is it being actively maintained and improved?

**Comment by rsmets** [comment]
@DaddyWarbucks apologies for missing your last couple of comments. Thanks for the `feathers-profiler` suggestion. 

In regards the parse-server, yes it has made a semi come back. The old startup that backed parse-server service was decommissioned after being purchased by facebook. However, the team was granted the ok the open source the framework. Turns out they had todo a total re-write so the open source version is not exactly the source that was powering the original parse-server service. However, it is super powerful and activity maintained by the community now. (All of the parse-server history is from memory... can't remember where I got all this info from ha or I would source it)
